doc/index.md

```diff
@@ -291,6 +291,103 @@ sentinel. That is only necessary if the underlying range's sentinel type models
+
+~~~~~~~{.cpp}
+    #include <range/v3/utility/basic_iterator.hpp>
+    #include <range/v3/utility/common_tuple.hpp>
+
+    using KeyIter   = typename std::vector<Key>::iterator;
+    using ValueIter = typename std::vector<Value>::iterator;
+
+    struct cursor {
+
+        // basic_iterator derives from "mixin", if present, so it can be used
+        // to inject things into the public interface of the iterator
+        struct mixin;
+
+        // This is for dereference operator.        
+        using value_type = ranges::common_pair<Key, Value>;
+        ranges::common_pair<Key&, Value&> read() const {
+            return { *key_iterator, *value_iterator };
+        }
+
+        bool equal(const cursor& other) const {
+            return key_iterator == other.key_iterator;
+        }
+
+        void next() { 
+            ++key_iterator;
+            ++value_iterator;
+        }
+
+        // prev optional. Required for Bidirectional iterator
+        void prev() { 
+            --key_iterator;
+            --value_iterator;
+        }
+
+        // advance and distance_to are optional. Required for RandomAcess iterator
+        void advance(std::ptrdiff_t n) { 
+            key_iterator   += n;
+            value_iterator += n;
+        }
+        std::ptrdiff_t distance_to(const cursor& other) const {
+            return other.key_iterator - this->key_iterator;
+        }
+
+        cursor() = default;
+        cursor(KeyIter key_iterator, ValueIter value_iterator)
+            : key_iterator(key_iterator)
+            , value_iterator(value_iterator)
+        {}
+
+        KeyIter   key_iterator;
+        ValueIter value_iterator;
+    };
+
+    struct cursor::mixin : ranges::basic_mixin<cursor>
+    {
+      using ranges::basic_mixin<cursor>::basic_mixin;
+
+      // It is necessary to expose constructor in this way
+      mixin(KeyIter key_iterator, ValueIter value_iterator)
+        : mixin{ cursor(key_iterator, value_iterator) }
+      {}
+
+      KeyIter key_iterator() {
+        return this->get().key_iterator;
+      }
+      ValueIter value_iterator() {
+        return this->get().value_iterator;
+      }
+    };    
+
+    using iterator = ranges::basic_iterator<cursor>;
+
+    void test(){
+      std::vector<Key>   keys   = {1};
+      std::vector<Value> values = {10};
+      
+      iterator iter(keys.begin(), values.begin());
+      auto   pair  = *iter;
+      Key&   key   = pair.first;
+      Value& value = pair.second;
+
+      auto key_iter = iter.key_iterator();
+    }
+~~~~~~~
+
+`read()` return references. So, we implicitly specify `value_type`.  
+ `ranges::common_pair` have conversions:  
+`ranges::common_pair<Key&, Value&>` <=> `ranges::common_pair<Key, Value>`  
+
+For more information, see [http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0186r0.html#basic-iterators-iterators.basic](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0186r0.html#basic-iterators-iterators.basic)
```

Comment:
Future-proof this link by changing it to <http://wg21.link/P0186#basic-iterators-iterators.basic>. If there is ever an updated version of P0186, this will resolve to that instead.